### ISSUE TITLE
Cannot update Wow or Overwatch on the btrfs volume with Battle.net client

### ISSUE THREAD

**Issue by SwimmingTiger** [maharmstone/btrfs#323]
Blizzard Battle.net cannot update World of Warcraft or Overwatch located on the btrfs volume. Every time you click update, it will end with the same error.

![2020-11-15 173544](https://user-images.githubusercontent.com/4986069/99181458-ae0bf280-2769-11eb-8d29-0453c068a32d.jpg)

The log in `AgentNGDP-20201115T084052.log`:

```
[I 2020-11-15 09:35:25.0045] - Update() starts with 0 items
[I 2020-11-15 09:35:25.0046] - Waiting the end condition...
[E 2020-11-15 09:35:25.0173] - failed to open manifest file 'D:/games/overwatch/_retail_/.flavor.info' for writing
[E 2020-11-15 09:35:25.0174] NGDP: Failed to write .flavor.info
[I 2020-11-15 09:35:25.0175] - Cancel updating...
```

Process Monitor v3.60 recorded a ACCESS DENIED.

![屏幕截图 2020-11-15 173613](https://user-images.githubusercontent.com/4986069/99181526-10fd8980-276a-11eb-98f8-cda28564c696.jpg)

But the current user obviously has read and write access to the file. In fact, it can be written normally when opened with any other editor (like notepad.exe).

Process Monitor v3.60 record file:

[Logfile.PML.zip](https://github.com/maharmstone/btrfs/files/5542131/Logfile.PML.zip)

--------------------------

WinBtrfs Version: 1.7.4, 1.7.5
Windows version: This problem exists in all Windows 10 versions I have used, from 1903 to 20H2.

--------------------------

## Notes:

* Not all Blizzard games will encounter update issues, Hearthstone can be updated correctly. I only found this problem with Wow and Overwatch in the Blizzard games I own.

   ![2020-11-15 175311](https://user-images.githubusercontent.com/4986069/99181765-a6e5e400-276b-11eb-91de-d3eb6abcf9a5.jpg)

* The game installation folder is not damaged. If you double-click the game exe directly, the game can be loaded normally.

   ![2020-11-15 175659](https://user-images.githubusercontent.com/4986069/99181809-fc21f580-276b-11eb-9826-3b69c4662a0a.jpg)

-------------------------------

I found a temporary solution to this issue. If I delete the `.flavor.info` file, Battle.net client will recreate it and the game update can be completed.

Of course, I need to manually delete the file every time when I update.

**Comment by maharmstone** [comment]
Thanks, I'll try to reproduce. I suspect this is a similar issue to https://github.com/maharmstone/btrfs/issues/291.

**Comment by Chaython** [comment]
Same as https://github.com/maharmstone/btrfs/issues/325 ? 
Though my drive is no longer writable at all. 
Read-Only is set to 0, I hadn't changed anything. Writing was working on Friday.
![image](https://user-images.githubusercontent.com/6486343/99929082-06626780-2d09-11eb-9b4a-c651bb3b261d.png)
![image](https://user-images.githubusercontent.com/6486343/99929093-17ab7400-2d09-11eb-98ba-c008e84e4450.png)

**Comment by maharmstone** [comment]
What does `btrfs check` say on Linux?

**Comment by YihaoPeng** [comment]
Just in case, I must remind:

**Running `btrfs check --repair` in Linux may cause a file system that has no major problems to become completely unmountable, and it cannot be repaired to a state where it can be mounted!**

I have encountered this problem before. My file system itself does not seem to have any problems, and `btrfs scrub` did not find any errors. But because of the issue of winbtrfs, I thought it had a problem, so I try to run `btrfs check --repair` to repair it.

However, the command exited with a segment fault! Then my file system became unmountable totally, both in Linux and Windows! I did not find any way to fix it so that it can be mounted again!

I had to use `btrfs restore` to restore the folders one by one, and found that I lost at least one folder's data.

**So just as `btrfs check --repair` warns: don't use it unless you have no choice!**

Of course, just running `btrfs check` (without `--repair`) will not cause file system damage, because it will not even perform any writes.

**Comment by maharmstone** [comment]
@YihaoPeng - erm, thanks, but nobody was talking about `btrfs check --repair`. And it does give a pretty unambiguous warning message if you try to use it.

> For my btrfs volume, it always reports several errors. But in fact, if you use `btrfs scrub`, you will find that the volume is completely fine.
They're measuring different things. Your volume is corrupted.

**Comment by YihaoPeng** [comment]
> They're measuring different things. Your volume is corrupted.

I do not think so. I later found that as long as I create a new volume, copy some files into it, and create some snapshots, `btrfs check` will report a few of errors.

So I think if the volume is normal in use, it is not damaged. The Linux btrfs driver developers have stated that the `btrfs check` code is not fully tested, and it is obviously more likely to have problems.

**Comment by YihaoPeng** [comment]
I found a temporary solution to this issue. If I delete the `.flavor.info` file, the Battle.net client will recreate it and the game update can be completed.

Of course, I need to manually delete the file every time when I update.

**Comment by maharmstone** [comment]
> > They're measuring different things. Your volume is corrupted.
> 
> I do not think so. I later found that as long as I create a new volume, copy some files into it, and create some snapshots, `btrfs check` will report a few of errors.
> 
> So I think if the volume is normal in use, it is not damaged. The Linux btrfs driver developers have stated that the `btrfs check` code is not fully tested, and it is obviously more likely to have problems.

I do know what I'm talking about, you know :-)
The last version of `btrfs check` with any false positives that I'm aware of was several years ago. Just because you never encounter any issues doesn't mean that your volume isn't subtly wrong - it might be that an inode you never access is corrupt, for instance.

**Comment by YihaoPeng** [comment]
> The last version of `btrfs check` with any false positives that I'm aware of was several years ago. Just because you never encounter any issues doesn't mean that your volume isn't subtly wrong - it might be that an inode you never access is corrupt, for instance.

Doesn't `btrfs scrub` access every inode in the file system?

If it will, why doesn't it find an error?

I think that if the code that actually runs the file system has self-healing capabilities, I should not use external tools (like `btrfs check`, it is not part of the file system code) to check it. The code that actually runs the file system will be responsible for it, while external tools may not: it may not be compatible with the latest developments in the file system.

**Comment by maharmstone** [comment]
You're confused about the difference between `scrub` and `check`.

Scrubbing verifies data and metadata against checksum errors, and restores it wherever possible. Scrubbing failures are nearly always caused by hard disk problems.

Checking checks that the volume is consistent - e.g. that the `st_nlink` value of an inode matches the number of hard links, that each directory entry is index both by number and by hash, that the extent tree matches what the inodes point to. Checking failures are only ever caused by driver problems, or possibly in rare cases by faulty RAM.

I've just noticed that you have the same profile pic as @SwimmingTiger, so I'm assuming you have two accounts. If you want me to look into your issue, please post the output of `btrfs check` as requested.

**Comment by YihaoPeng** [comment]
> I've just noticed that you have the same profile pic as @SwimmingTiger, so I'm assuming you have two accounts. If you want me to look into your issue, please post the output of `btrfs check` as requested.

Thank you. Both are my account.

But I think I should really submit an issue to the Linux `btrfs` utility.

I really don't think there is a problem with my volume because it looks so normal. And even if I re-created a volume, a similar error was reported quickly.

**Comment by maharmstone** [comment]
Closing as user refused to give information requested.

**Comment by YihaoPeng** [comment]
The issue has been closed, but as a supplement, I don’t think the result of `btrfs check` is related to the issue.

Because actually, when my volume was completely damaged by `btrfs check --repair`, I reformatted the volume and copied the game files again in Linux. But the result is that the same problem still exists on Windows.

So, this should be a issue about how to deal with Windows file permissions (ACL), not a issue about volume damage.

And, when I run the Battle.net client with Wine on Linux, everything is fine. This again proves that the volume is not damaged.

The reason why I mentioned the results of my `btrfs check` is to prevent others from damaging their volumes by running `btrfs check --repair`. Unfortunately, the subsequent discussion completely deviated from the direction.
